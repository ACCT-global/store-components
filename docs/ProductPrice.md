ðŸ“¢ Don't fork this project. Use, [contribute](https://github.com/vtex-apps/awesome-io#contributing), or open issues through [Store Discussion](https://github.com/vtex-apps/store-discussion).

# Product Price

`ProductPrice` is a VTEX component that shows the price of a product.
This Component can be imported and used by any VTEX app.

## Configuration

1. Import the vtex.store-component's app to your theme's dependencies in the manifest.json, for example:

```json
  dependencies: {
    "vtex.store-components": "3.x"
  }
```

2. Add the `product-price` block to any block bellow `store.product`. For example:

```json
  "store.product": {
    "children": [
      "flex-layout.row#product",
    ]
  },
  "flex-layout.row#product": {
    "children": [
      "product-price"
    ]
  }
```

| Prop name               | Type      | Description                           | Default value |
| ----------------------- | --------- | ------------------------------------- | ------------- |
| `labelSellingPrice`     | `String`  | Product selling price label           | null          |
| `labelListPrice`        | `String`  | Product list price label              | null          |
| `showListPrice`         | `Boolean` | Set visibility of list price          | true          |
| `sellingPrices`         | `Array`   | Product list of selling prices        | []            |
| `showSellingPriceRange` | `Boolean` | Set visibility of selling price range | false         |
| `showListPriceRange`    | `Boolean` | Set visibility of list price range    | false         |
| `showLabels`            | `Boolean` | Set visibility of labels              | true          |
| `showInstallments`      | `Boolean` | Set visibility of installments        | false         |
| `showSavings`           | `Boolean` | Set visibility of savings             | false         |

## Customization

| CSS Handles                 |
| -------------------------- |
| `price_className` |
| `price_loader` |
| `price_listPriceContainer` |
| `price_listPriceLabel` |
| `price_listPrice` |
| `price_listPriceRange` |
| `price_sellingPriceRange` |
| `price_sellingPriceContainer` |
| `price_sellingPriceLabel` |
| `price_sellingPrice` |
| `price_savingsContainer` |
| `price_savings` |
| `price_installment` |
| `price_interestRate` |
| `price_installmentContainer` |
