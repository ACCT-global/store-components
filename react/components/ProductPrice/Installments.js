import React, { Component, Fragment } from 'react'
import classNames from 'classnames'
import { isEmpty } from 'ramda'
import { FormattedMessage } from 'react-intl'
import PropTypes from 'prop-types'

import PricePropTypes from './propTypes'

/** Installments component */
export default class Installments extends Component {

  static defaultProps = {
    classes: {
      root: null,
      installmentValue: null,
      interestRate: null
    }

  }

  static propTypes = {
    /** Product installments to be displayed */
    installments: PricePropTypes.installments,
    /** Pages editor config to display labels */
    showLabels: PropTypes.bool.isRequired,
    /** react-intl function to format the prices*/
    formatNumber: PropTypes.func.isRequired,
    /** Options to be passe to the formatNumber function*/
    currencyOptions: PropTypes.shape({
      style: PropTypes.string.isRequired,
      currency: PropTypes.string.isRequired,
      minimumFractionDigits: PropTypes.number.isRequired,
      maximumFractionDigits: PropTypes.number.isRequired,
    }).isRequired,
    /** Classes to be applied in the Installments Component */
    classes: PropTypes.shape({
      root: PropTypes.string,
      installmentValue: PropTypes.string,
      interestRate: PropTypes.string
    })
  }
  
  render() {
    const {
      classes,
      showLabels,
      formatNumber,
      installments,
      currencyOptions,
    } = this.props

    if (!installments || isEmpty(installments)) {
      return null
    }

    const noInterestRateInstallments = installments.filter(
      installment => !installment.InterestRate
    )

    /*
     * - The selected installment will be the one with the highest `NumberOfInstallments`;
     * - If there is no 'interest-free' installments, the normal installments will be analyzed.
     */
    const installment = (isEmpty(noInterestRateInstallments)
      ? installments
      : noInterestRateInstallments
    ).reduce(
      (previous, current) =>
        previous.NumberOfInstallments > current.NumberOfInstallments
          ? previous
          : current
    )

    const formattedInstallmentPrice = formatNumber(
      installment.Value,
      currencyOptions
    )

    const [installmentsElement, installmentPriceElement, timesElement] = [
      installment.NumberOfInstallments,
      formattedInstallmentPrice,
      <Fragment>&times;</Fragment>,
    ].map((element, index) => (
      <span className={classes.installmentValue} key={index}>
        {element}
      </span>
    ))

    return (
      <div className={classNames('vtex-price-installments__container', classes.root)}>
        {showLabels ? (
          <FormattedMessage
            id="pricing.installment-display"
            values={{
              installments: installmentsElement,
              installmentPrice: installmentPriceElement,
              times: timesElement,
            }}
          />
        ) : (
            <Fragment>
              {installmentsElement} {timesElement} {installmentPriceElement}
            </Fragment>
          )}
        {!installment.InterestRate && (
          <div className={classNames('vtex-price-installments__interest-rate', classes.interestRate)}>
            <FormattedMessage id="pricing.interest-free" />
          </div>
        )}
      </div>
    )
  }
}
