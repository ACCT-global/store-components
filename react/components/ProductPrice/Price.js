import React from 'react'
import { map, join } from 'ramda'
import { injectIntl } from 'react-intl'

const isValidPriceRange = priceRange => {
  const [lowPrice, highPrice] = priceRange
  return priceRange.length === 2 && lowPrice !== highPrice
}

const Price = ({
  showPriceRange,
  priceRange,
  price,
  rangeContainerClasses,
  singleContainerClasses,
  intl: { formatNumber },
  currencyOptions,
}) => {
  const mustShowPriceRange = showPriceRange && priceRange && isValidPriceRange(priceRange)

  const formatPrice = rawPrice => {
    return formatNumber(rawPrice, currencyOptions)
  }

  const formatPriceRange = rawPriceRange => {
    const priceRangeFormatted = map(formatPrice, rawPriceRange || [])
    return join(' - ', priceRangeFormatted)
  }

  if (mustShowPriceRange) {
    return (
      <span className={rangeContainerClasses}>
        {formatPriceRange(priceRange)}
      </span>
    )
  }
  return (
    <span className={singleContainerClasses}>
      {formatPrice(price)}
    </span>
  )
}

export default injectIntl(Price)
