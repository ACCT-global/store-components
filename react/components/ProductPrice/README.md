# Product Price

## Description

`ProductPrice` is a VTEX component that shows the price of a product.
This Component can be imported and used by any VTEX app.

:loudspeaker: **Disclaimer:** Don't fork this project, use, contribute, or open issue with your feature request.

## Table of Contents

## Table of Contents
- [Usage](#usage)
  - [Blocks API](#blocks-api)
  - [Configuration](#configuration)
  - [Styles API](#styles-api)

## Usage

You should follow the usage instruction in the main [README](https://github.com/vtex-apps/store-components/blob/master/README.md#usage).

Then, add `product-price` block into your app theme, as we do in our [Product Details app](https://github.com/vtex-apps/product-details/blob/master/store/blocks.json). 

### Blocks API

This component has an interface that describes which rules must be implemented by a block when you want to use the `ProductPrice`.

 ```json
  "product-price": {
    "component": "ProductPrice"
  }
```	

### Configuration

Through the Storefront, you can change the `ProductPrice`'s behavior and interface. However, you also can make in your theme app, as Dreamstore does.

| Prop name | Type | Description |
| --------- | ---- | ----------- |
| `sellingPrice` | `Number` | Product selling price |
| `listPrice` | `Number` | Product list price |
| `showListPrice` | `Boolean!` | Set visibility of list price |
| `showLabels` | `Boolean!` | Set visibility of labels |
| `showInstallments` | `Boolean!` | Set visibility of installments |
| `showSavings` | `Boolean` | Set visibility of savings |
| `installments` | `Array` | Available installments |
| `installments.Value` | `Number!` | Installment price |
| `installments.InterestRate` | `Number!` | Installment interest rate (zero if interest-free) |
| `installments.TotalValuePlusInterestRate` | `Number` | Calculated total value |
| `installments.NumberOfInstallments` | `Number!` | Number of installments |
| `installments.Name` | `String` | Installment offer name |
| `className` | `String` | Classes to be applied to root element |
| `loaderClass` | `String` | Classes to be applied to loader root element |
| `listPriceContainerClass` | `String` | Classes to be applied to container of list price |
| `listPriceLabelClass` | `String` | Classes to be applied to label of price |
| `listPriceClass` | `String` | Classes to be applied to price value |
| `sellingPriceContainerClass` | `String` | Classes to be applied to selling price container |
| `sellingPriceLabelClass` | `String` | Classes to be applied to selling price label |
| `sellingPriceClass` | `String` | Classes to be applied to selling price value |
| `savingsContainerClass` | `String` | Classes to be applied to savings container |
| `savingsClass` | `String` | Classes to be applied to savings |
| `installmentClass` | `String` | Classes to be applied to installment element |
| `installmentContainerClass` | `String` | Classes to be applied to installment container |
| `interestRateClass` | `String` | Classes to be applied to interest rate element |

### Styles API
You should follow the Styles API instruction in the main [README](https://github.com/vtex-apps/store-components/blob/master/README.md#styles-api).

| Token name | Component | Description |
| ---------- | --------- | ----------- |