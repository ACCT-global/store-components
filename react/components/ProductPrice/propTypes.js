import PropTypes from 'prop-types'
import { intlShape } from 'react-intl'

export default {
  /** Product selling price */
  sellingPrice: PropTypes.number,
  /** Product list price */
  listPrice: PropTypes.number,
  /** Set visibility of list price */
  showListPrice: PropTypes.bool.isRequired,
  /** Set visibility of labels */
  showLabels: PropTypes.bool.isRequired,
  /** Set visibility of installments */
  showInstallments: PropTypes.bool.isRequired,
  /** Set visibility of savings */
  showSavings: PropTypes.bool,
  /** Text to the selling price's label */
  labelSellingPrice: PropTypes.string,
  /** Available installments */
  installments: PropTypes.arrayOf(
    PropTypes.shape({
      /** Installment value */
      Value: PropTypes.number.isRequired,
      /** Interest rate (zero if interest-free) */
      InterestRate: PropTypes.number.isRequired,
      /** Calculated total value */
      TotalValuePlusInterestRate: PropTypes.number,
      /** Number of installments */
      NumberOfInstallments: PropTypes.number.isRequired,
      /** Installment offer name */
      Name: PropTypes.string,
    })
  ),
  /** CSS classes to be applied in the Price component Component */
  classes: PropTypes.shape({
    root: PropTypes.string,
    rootLoader: PropTypes.string,
    listPrice: PropTypes.shape({
      container: PropTypes.string,
      label: PropTypes.string,
      value: PropTypes.string
    }),
    sellingPrice: PropTypes.shape({
      container: PropTypes.string,
      label: PropTypes.string,
      value: PropTypes.string
    }),
    savings: PropTypes.shape({
      container: PropTypes.string,
      value: PropTypes.string
    })
  }),
  /** Classes to be applied in the Installments Component */
  installmentsClasses: PropTypes.shape({
    root: PropTypes.string,
    installmentValue: PropTypes.string,
    interestRate: PropTypes.string
  }),
  /** Component and content loader styles */
  styles: PropTypes.object,
  /** intl property to format data */
  intl: intlShape.isRequired,
}
