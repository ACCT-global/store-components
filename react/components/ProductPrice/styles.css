/**
*   ProductPrice CSS Modules
*/
.priceContainer {}
.listPrice {}
.listPriceLabel {}
.listPriceValue {}

.sellingPrice {}
.sellingPriceLabel {}
.sellingPriceValue {}

.savingPrice {}
.savingPriceValue {}

.installmentsPrice {}
.interestRatePrice {}

.priceLoaderContainer {}
.listPriceLoader {}
.sellingPriceLabelLoader {}
.sellingPriceLoader {}
.installmentsPriceLoader {}
.savingsPriceLoader {}