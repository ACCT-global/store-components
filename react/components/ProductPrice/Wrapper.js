import * as React from 'react'
import { ProductContext } from 'vtex.product-context'
import { path, isEmpty } from 'ramda'

import ProductPrice from './index'

const styles = {
  'vtex-price-list__container--loader': {
    x: 0,
    width: '7.219em',
    height: '0.56em',
  },
  'vtex-price-selling__label--loader': {
    x: 0,
    y: '2em',
    width: '2.85em',
    height: '1.08em',
  },
  'vtex-price-selling--loader': {
    x: '3.25em',
    y: '0.86em',
    width: '14.572em',
    height: '2.176em',
  },
  'vtex-price-installments--loader': {
    x: 0,
    y: '3.75em',
    width: '12em',
    height: '0.825em',
  },
  'vtex-price-savings--loader': {
    x: 0,
    y: '5em',
    width: '10em',
    height: '0.686em',
  },
}

const ProductPriceWrapper = ({
  labelSellingPrice,
  showListPrice,
  showInstallments,
  showLabels,
  showSavings,
  ...props,
}) => {
  const valuesFromContext = React.useContext(ProductContext)

  const productPriceProps = () => {
    if (!valuesFromContext || isEmpty(valuesFromContext)) 
      return {
        ...props,
        labelSellingPrice,
        showListPrice,
        showInstallments,
        showLabels,
        showSavings,      
      }

    const { selectedItem } = valuesFromContext
    const commertialOffer = path(['sellers', 0, 'commertialOffer'], selectedItem)

    return {
      styles: styles,
      listPriceContainerClass: 't-small-s t-small-ns c-muted-2 mb2',
      sellingPriceLabelClass: 't-heading-6-s t-heading-5-ns dib',
      listPriceLabelClass: 'dib strike',
      listPriceClass: 'ph2 dib strike',
      sellingPriceContainerClass: 'pv1 b c-on-base',
      sellingPriceClass: 't-heading-2-s dib ph2',
      installmentContainerClass: 't-mini-s t-small-ns c-on-base',
      installmentClass: 't-body',
      interestRateClass: 'dib ph2',
      savingsContainerClass: 'c-success mt3',
      savingsClass: 'dib t-small',
      loaderClass: 'h4-s mw6-s pt2-s',
      listPrice: path(['ListPrice'], commertialOffer),
      sellingPrice: path(['Price'], commertialOffer),
      installments: path(['Installments'], commertialOffer),
      labelSellingPrice,
      showLabels,
      showInstallments,
      showListPrice,
      showSavings,    
    }
  }

  return (
    <ProductPrice { ...productPriceProps() } />
  )
}

ProductPriceWrapper.schema = {
  title: 'admin/editor.productPrice.title',
  description: 'admin/editor.productPrice.description',
  type: 'object',
  properties: {
    labelSellingPrice: {
      type: 'string',
      title: 'admin/editor.productPrice.labelSellingPrice',
      default: null,
      isLayout: true,
    },
    showListPrice: {
      type: 'boolean',
      title: 'admin/editor.productPrice.showListPrice',
      default: true,
      isLayout: true,
    },
    showLabels: {
      type: 'boolean',
      title: 'admin/editor.productPrice.showLabels',
      default: true,
      isLayout: true,
    },
    showInstallments: {
      type: 'boolean',
      title: 'admin/editor.productPrice.showInstallments',
      default: false,
      isLayout: true,
    },
    showSavings: {
      type: 'boolean',
      title: 'admin/editor.productPrice.showSavings',
      default: false,
      isLayout: true,
    },
  },
}

export default ProductPriceWrapper
