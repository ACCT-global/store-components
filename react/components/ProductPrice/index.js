import React, { Component } from 'react'
import classNames from 'classnames'
import PropTypes from 'prop-types'
import { isNil } from 'ramda'
import ContentLoader from 'react-content-loader'
import { FormattedMessage, injectIntl } from 'react-intl'

import PricePropTypes from './propTypes'
import Installments from './Installments'

import productPrice from './styles.css'
/**
 * The Price component. Shows the prices information of the Product Summary.
 */
class Price extends Component {
  static contextTypes = {
    culture: PropTypes.object,
  }

  static propTypes = PricePropTypes

  static defaultProps = {
    showListPrice: true,
    showLabels: true,
    showInstallments: false,
    showSavings: false,
    labelSellingPrice: null,
  }

  currencyOptions = {
    style: 'currency',
    currency: this.context.culture.currency,
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  }

  render() {
    const {
      sellingPrice,
      listPrice,
      showListPrice,
      showInstallments,
      showLabels,
      showSavings,
      labelSellingPrice,
      className,
      loaderClass,
      listPriceContainerClass,
      listPriceLabelClass,
      listPriceClass,
      sellingPriceContainerClass,
      sellingPriceLabelClass,
      sellingPriceClass,
      savingsContainerClass,
      savingsClass,
      installments,
      installmentClass,
      interestRateClass,
      installmentContainerClass,
      styles,
      intl: { formatNumber },
    } = this.props

    let { classes } = this.props
    // avoiding undefined verifications
    classes = {
      ...Price.defaultProps.classes,
      ...classes,
    }

    if ((showListPrice && isNil(listPrice)) || isNil(sellingPrice)) {
      return <Price.Loader loaderClass={loaderClass} {...styles} />
    }

    const differentPrices = showListPrice && sellingPrice !== listPrice

    return (
      <div className={classNames(productPrice.priceContainer, className)}>
        {differentPrices && (
          <div
            className={classNames(
              productPrice.listPrice,
              listPriceContainerClass
            )}
          >
            {showLabels && (
              <div
                className={classNames(
                  productPrice.listPriceLabel,
                  'dib ph2 t-small-ns t-mini'
                )}
              >
                <FormattedMessage id="store/pricing.from" />
              </div>
            )}
            <span
              className={classNames(
                productPrice.listPriceValue,
                listPriceClass
              )}
            >
              {formatNumber(listPrice, this.currencyOptions)}
            </span>
          </div>
        )}
        <div
          className={classNames(
            productPrice.sellingPrice,
            sellingPriceContainerClass
          )}
        >
          {showLabels && listPrice !== sellingPrice && (
            <div
              className={classNames(
                productPrice.sellingPriceLabel,
                sellingPriceLabelClass
              )}
            >
              {labelSellingPrice || <FormattedMessage id="store/pricing.to" />}
            </div>
          )}
          <div
            className={classNames(productPrice.sellingPrice, sellingPriceClass)}
          >
            {formatNumber(sellingPrice, this.currencyOptions)}
          </div>
        </div>
        {showInstallments && (
          <Installments
            installments={installments}
            showLabels={showLabels}
            formatNumber={formatNumber}
            currencyOptions={this.currencyOptions}
            className={installmentContainerClass}
            interestRateClass={interestRateClass}
            installmentClass={installmentClass}
          />
        )}
        {differentPrices && showSavings && (
          <div
            className={classNames(
              productPrice.savingPrice,
              savingsContainerClass
            )}
          >
            <div
              className={classNames(
                productPrice.savingPriceValue,
                savingsClass
              )}
            >
              <FormattedMessage
                id="store/pricing.savings"
                values={{
                  savings: formatNumber(
                    listPrice - sellingPrice,
                    this.currencyOptions
                  ),
                }}
              />
            </div>
          </div>
        )}
      </div>
    )
  }
}

Price.Loader = (loaderProps = {}) => (
  <div
    className={classNames(
      productPrice.priceContainer,
      productPrice.priceLoaderContainer,
      loaderProps.loaderClass
    )}
  >
    <ContentLoader
      style={{
        width: '100%',
        height: '100%',
      }}
      width={300}
      height={70}
      preserveAspectRatio="xMinYMin meet"
      {...loaderProps}
    >
      <rect
        height="0.75em"
        width="50%"
        x="25%"
        {...loaderProps[productPrice.listPriceLoader]}
      />
      <rect {...loaderProps[productPrice.sellingPriceLabelLoader]} />
      <rect
        height="1em"
        width="70%"
        x="15%"
        y="1.25em"
        {...loaderProps[productPrice.sellingPriceLoader]}
      />
      <rect
        height="0.75em"
        width="80%"
        x="10%"
        y="2.75em"
        {...loaderProps[productPrice.installmentsPriceLoader]}
      />
      <rect {...loaderProps[productPrice.savingsPriceLoader]} />
    </ContentLoader>
  </div>
)

Price.Loader.displayName = 'Price.Loader'

export default injectIntl(Price)
