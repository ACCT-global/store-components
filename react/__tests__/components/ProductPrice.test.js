import React from 'react'
import ProductPrice from './../../ProductPrice'
import { mount } from 'enzyme'

describe("<ProductPrice />", () => {
  let renderComponent = null
  const defaultProps = {
    loaderClass: '',
    intl: { formatNumber: (number, currency) => `${currency} ${number}` }
  }
  
  const context = { culture: { currency : 'USD' } };

  beforeEach(() => {
    renderComponent = customProps => {
      const props = {
        ...defaultProps,
        ...customProps,
      }
      return mount(<ProductPrice {...props} />, { context })
    }
  })

  it('should be mount', () => {
    expect(renderComponent()).toBeDefined()
  })

  it('should match snapshot', () => {
    expect(renderComponent()).toMatchSnapshot()
  })
})