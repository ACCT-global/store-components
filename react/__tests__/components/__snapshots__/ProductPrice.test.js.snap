// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductPrice /> should match snapshot 1`] = `
<Price
  intl={
    Object {
      "formatNumber": [Function],
    }
  }
  labelSellingPrice={null}
  loaderClass=""
  showInstallments={false}
  showLabels={true}
  showListPrice={true}
  showSavings={false}
>
  <Price.Loader
    loaderClass=""
  >
    <div
      className="priceContainer priceLoaderContainer"
    >
      <ContentLoad
        height={70}
        loaderClass=""
        preserveAspectRatio="xMinYMin meet"
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
        width={300}
      >
        <div>
          <rect
            height="0.75em"
            width="50%"
            x="25%"
          />
          <rect />
          <rect
            height="1em"
            width="70%"
            x="15%"
            y="1.25em"
          />
          <rect
            height="0.75em"
            width="80%"
            x="10%"
            y="2.75em"
          />
          <rect />
        </div>
      </ContentLoad>
    </div>
  </Price.Loader>
</Price>
`;
